### DISCUSSION TITLE
Move opacity custom properties to universal selector

### DISCUSSION THREAD

**Discussion by MartijnCuppens** [tailwindlabs/tailwindcss#4603] [Ideas]
For the non-JIT mode, every `.text-opacity-*`, `.bg-opacity-*`, `.border-opacity-*`, `.ring-opacity-*`, and `divide-opacity-*` add custom properties (`--tw-*-opacity: 1;`) to their selectors. This could be simplified by moving the default values to the universal selector.

It would also be a little more consistent since custom properties like `--tw-shadow` are already in the universal selector.

JIT solves this with https://github.com/tailwindlabs/tailwindcss/pull/4348, but the primary reason I don't use the JIT-mode is because I use custom properties, because my colors change per page.

**Comment by adamwathan** [comment]
Hey man! The only issue I see with this is that it's a breaking change for the following example:

```html
<div class="bg-red-500 bg-opacity-50 md:bg-blue-500">
```

Currently, on `md` screens that will jump back to 100% opacity. You could definitely make the argument that it should still be 50% opacity and maybe that actually is more intuitive, but I get a bit afraid of breaking changes like this because they can be really hard for people to update in their projects (imagine writing a regex or something to detect "instances where I used an opacity utility, then added a new color utility on hover or at a different breakpoint but didn't re-specify the opacity utility" 🥴).

I would consider it for sure if we knew it would be a low impact change across the community though.

> the primary reason I don't use the JIT-mode is because I use custom properties, because my colors change per page.

I'm curious why this would prevent you from using JIT since the generated CSS is the same either way? Would love to hear about a situation where it wouldn't work.

**Reply by MartijnCuppens** [reply]
> but I get a bit afraid of breaking changes like this because they can be really hard for people to update in their projects

Yup, that's definitely a BC, didn't think of that.

> I'm curious why this would prevent you from using JIT since the generated CSS is the same either way? Would love to hear about a situation where it wouldn't work.

I clearly underestimated the flexibility of the JIT mode, just tested it and indeed custom property support just works out of the box. That's some pretty amazing stuff you made there 👏
